### ISSUE TITLE
False positive needless_pass_by_value in procedural macros

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#3067]
```ini
[lib]
proc-macro = true
```

```rust
extern crate proc_macro;
use proc_macro::TokenStream;

#[proc_macro]
pub fn my_macro(input: TokenStream) -> TokenStream {
    println!("{}", input);
    TokenStream::new()
}
```

Clippy flags the following:

```console
warning: this argument is passed by value, but not consumed in the function body
 --> src/lib.rs:5:24
  |
5 | pub fn my_macro(input: TokenStream) -> TokenStream {
  |                        ^^^^^^^^^^^ help: consider taking a reference instead: `&TokenStream`
  |
  = note: #[warn(needless_pass_by_value)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value
```

The suggestion is not correct because the compiler requires the signature of a procedural macro to accept a TokenStream by value.

```console
error[E0308]: mismatched types
 --> src/lib.rs:5:1
  |
5 | / pub fn my_macro(input: &TokenStream) -> TokenStream {
6 | |     println!("{}", input);
7 | |     TokenStream::new()
8 | | }
  | |_^ expected struct `proc_macro::TokenStream`, found reference
  |
  = note: expected type `fn(proc_macro::TokenStream) -> proc_macro::TokenStream`
             found type `for<'r> fn(&'r proc_macro::TokenStream) -> proc_macro::TokenStream {my_macro}`
```

**Comment by dtolnay** [comment]
Same thing for `#[proc_macro_attribute]` macros.

**Comment by mcarton** [comment]
Closed by #1617.
